core_types/src/agent_id.rs

```diff
@@ -0,0 +1,156 @@
+//! Agent Ids are 64 bytes long.
+//! - The first 32 bytes 0-32 are the public side of a libsodium signing keypair
+//! - The next 32 bytes 32-64 are the public side of a libsodium kx keypair
+//! When displayed to users 2 reed-solomon parity bytes are appended to the end,
+//! and the resulting buffer is base64url encoded.
+//! (https://tools.ietf.org/html/rfc4648#section-5)
+
+use std::convert::TryFrom;
+
+use super::error::HolochainError;
+use reed_solomon::{Decoder, Encoder};
+
+const PARITY_LEN: usize = 2;
+
+/// Represents a public agent identity
+pub struct AgentIdPub([u8; 64]);
+
+impl AgentIdPub {
+    /// get the signing public key portion of this identity
+    pub fn get_sig(&self) -> &[u8; 32] {
+        array_ref![self.0, 0, 32]
+    }
+
+    /// get the encryption (kx) public key portion of this identity
+    pub fn get_enc(&self) -> &[u8; 32] {
+        array_ref![self.0, 32, 32]
+    }
+}
+
+impl std::fmt::Display for AgentIdPub {
+    /// display this AgentIdPub as a user string
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{}", String::from(self))
+    }
+}
+
+impl std::fmt::Debug for AgentIdPub {
+    /// output the raw bytes of this public identity
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "AgentIdPub(sig({:?}), enc({:?}))",
+            self.get_sig(),
+            self.get_enc()
+        )
+    }
+}
+
+impl<'a> From<&'a AgentIdPub> for String {
+    /// stringify this agent public identity
+    fn from(id: &AgentIdPub) -> Self {
+        let enc = Encoder::new(PARITY_LEN);
+        let enc = *enc.encode(&id.0[0..64]);
+        base64::encode(&enc[..]).replace("+", "-").replace("/", "_")
```

Comment:
What is this replacement for?

Reply:
The [base64url variant](https://tools.ietf.org/html/rfc4648#section-5) makes the ids easier to use in filenames and urls, (no forward slash).